modules/substrate/src/lib.rs

```diff
@@ -376,19 +376,7 @@ impl<T: Config> bp_header_chain::HeaderChain<BridgedHeader<T>> for Module<T> {
 			storage: PalletStorage::<T>::new(),
 		};
 
-		let _ = verifier.import_header(header.hash(), header).map_err(|_| ())?;
-
-		Ok(())
-	}
-
-	fn import_finality_proof(header: BridgedHeader<T>, finality_proof: Vec<u8>) -> Result<(), ()> {
-		let mut verifier = verifier::Verifier {
-			storage: PalletStorage::<T>::new(),
-		};
-
-		let _ = verifier
-			.import_finality_proof(header.hash(), finality_proof.into())
-			.map_err(|_| ())?;
+		let _ = verifier.import_header_unchecked(header).map_err(|_| ())?;
```

Comment:
I'm not sure about changing the existing trait. Why not add a new one that simply adds a different interface towards the same `Module`?

I.e. we keep `import_header/import_finality_proof` and have a more specific `trait FinalizedHeaderChain: HeaderChain` which only exposes a unchecked function to import these pre-verified headers.

Reply:
I don't think we should be exposing both `import_header/import_finality` and `import_unchecked` because for our purposes `import_unchecked` is superior to the former methods.

Unless you can convince me that having all three methods is useful, I'm going to only work with the unchecked method 😄

Reply:
I feel that redesigning (& merging) both pallets will be needed in the future, and by changing `HeaderChain`, we are risking leaving some incorrect documentation behind (or breaking some assumptions in other parts of the code).
The fact that `HeaderChain::import_header` method stays the same on the API level, but expects a completely different thing (a fully-verified, final header that builds on top of the best finalized header) worries me.
Also bear in mind, that currently the code is fully functional, and as mentioned earlier I think we should rather do a complete overhaul in the future, then partially change existing things given they are used in a different context.
The ongoing audit is another reason - the main point of introducing new pallet was to avoid touching existing code, and I'd rather keep it this way.

I know none of these arguments is actually really strong one, most of them are just based on gut feelings, but perhaps that's enough to convince you :)

Reply:
Okay so I think there's a small misunderstanding here. The `HeaderChain` trait was introduced  and implemented on `substrate_bridge_pallet::Module` as part of #629. The methods it exposes are not used by any old code (only the Finality Verifier pallet calls them). The relayer uses the `import_signed_header` and `finalize_header` dispatchables. Changing `HeaderChain` and its implementation doesn't affect any exiting functionality.

Reply:
>  Changing HeaderChain and its implementation doesn't affect any exiting functionality.

...most likely.

The same argument applies for the header pallet - since it's not used in any bridge we could freely change it, but again my point is to minimize changes to avoid making some weird mistakes - we shouldn't need to carefully review if the code is used or not, or if the assumptions are not getting broken by introducing this change or another.
But again, that's just my gut feeling :)

Reply:
Take another look at the updated code and let me know if this is still bugging you